core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java

```diff
@@ -192,6 +192,14 @@ public Table create() {
 
       String baseLocation = location != null ? location : defaultWarehouseLocation(identifier);
       tableProperties.putAll(tableOverrideProperties());
+
+      if (Boolean.parseBoolean(tableProperties.get(TableProperties.UNIQUE_LOCATION))) {
+        boolean alreadyExists = ops.io().newInputFile(baseLocation).exists();
+        if (alreadyExists) {
+          throw new AlreadyExistsException("Table location already in use: %s", baseLocation);
```

Comment:
Not quite accurate since we are just checking whether the directory is there, which is different than the location being in use. While I think this may be good to check, we've seen most errors here being users setting table locations as sub directories as other tables. 

I think maybe we should check that the directory/location is empty? That's probably a not too aggressive check?

Reply:
thank you @RussellSpitzer for your feedback. I totally agree that checking for directory or location is empty would be awesome, but based on the methods available in [FileIO](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/io/FileIO.java). I dont find anything useful for that purpose. The closest one I get is [SupportsPrefixOperations::listPrefix](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/io/SupportsPrefixOperations.java#L36) which is subclass of fileIO. 

Do you think it make sense for us to do conditional check based on io?

```java
boolean alreadyExists =
  (ops.io() instanceof SupportsPrefixOperations) ?
      Iterables.isEmpty(
          ((SupportsPrefixOperations) ops.io()).listPrefix(baseLocation)
      )
      : ops.io().newInputFile(baseLocation).exists();
```

Reply:
I have thought through this and mostly two cases came to my mind. We may think with this route

1. No database creation should be allowed under an existing database path. It will help a major problem of people creating even databases under existing db path. 
2. No table creation should be allowed under an existing table path. 

**Case 1**

We have the following information with us which is an existing Table and it's location. 

Say a table's location is `s3://somerandompath/my_database/my_table `

If we make it much simple and try to set the reference path of a table to database level that can help may be. 
We have various ways of creating iceberg table just via database.tableName, with location etc. This DB path is always a constant path by practice. If someone is trying to create a table under the same location with same name we can just throw the exception that s3://somerandompath/my_database/my_table exists just by looking it's database reference. The uniqueness not necessarily you need from storage file location

```
CREATE TABLE prod.db.sample
USING iceberg
PARTITIONED BY (part)
TBLPROPERTIES ('key'='value')
AS SELECT ...
```

OR 

```
CREATE TABLE IF NOT EXISTS prod.db.sample (
         id integer,
        ......
       )
       USING ICEBERG 
       LOCATION 
       TBLPROPERTIES (
         'type' 'hive',.......
       )
```


**Case 2**

Rename table : When we rename a table we don't move files it is a metadata operation. The base path remains same but the table name gets updated. So in this case there is no impact. For unique location we can still look up to the metadata and get all unique paths under db reference.